euphony/src/main/cpp/core/charset/CharsetAutoSelector.cpp

```diff
@@ -0,0 +1,29 @@
+#include "CharsetAutoSelector.h"
+#include "ASCIICharset.h"
+#include "DefaultCharset.h"
+#include "UTF8Charset.h"
+#include "UTF16Charset.h"
+#include "UTF32Charset.h"
+#include <sstream>
+#include <iomanip>
+#include <iostream>
+#include <algorithm>
+
+using namespace Euphony;
+
+HexVector CharsetAutoSelector::select(std::string src) {
+
+    HexVector asciiResult = ASCIICharset().encode(src);
+    HexVector defaultResult = DefaultCharset().encode(src);
+    HexVector utf8Result = UTF8Charset().encode(src);
+    HexVector utf16Result = UTF16Charset().encode(src);
+    HexVector utf32Result = UTF32Charset().encode(src);
+
+    HexVector results[] = {asciiResult, defaultResult, utf8Result, utf16Result, utf32Result};
+
+    std::sort(results, results+5, [](HexVector& left, HexVector& right) {
+        return left.getSize() < right.getSize();
+    });
+    
+    return results[0];
+}
```

Comment:
Good job to make `select` function
It'll make a shortest encoded data from `sort`.

Seeing this result, I thought it would be good to implement the Charset interface called AutoCharset.
```cpp
class AutoCharset : public Charset {
  // encode & decode implemenation
}
```

However, what I expected is that `select` is gonna tell us type of charset.
this is utility for charset selection. so, function definition might be like below.
your proposal is also this one https://github.com/euphony-io/euphony/issues/222

```cpp
Chraset* ChrasetAutoSelector::select(std::string src);
```
or
```cpp
shared_ptr<Charset> CharsetAutoSelector::select(std::string src);
```

How do you think about it?

Reply:
I think so. It has been revised and I will commit soon.

Reply:
@DahyeonWoo 
Here is my thinking.
There are so many methods to implement `CharsetAutoSelector` :)

Firstly, analyze the charset's charecteristics.
- ASCII is excluded because UTF-8 is included
- UTF32 is unconditionally 4 bytes, except that it always returns the maximum length.
- All that is left is to compare UTF-8 and UTF-16
- If there are only Korean characters, UTF16 is short (2 bytes per Korean character)
- If there is a lot of English and sometimes a little bit of Korean, UTF8 is short (because English is 1 byte and Korean is 3 bytes)

My pseudo code :)
```cpp
std::shared_ptr<Charset> CharsetAutoSelector::select(std::string src) {
    std::shared_ptr<Charset> utf8Charset = make_shared<UTF8Charset>();
    std::shared_ptr<Charset> utf16Charset = make_shared<UTF16Charset>();

    return (utf8Charset->encode(src) > utf16Charset->encode(src) ? utc16Charset : utf8Charset;
}
```

Reply:
And.. if return type is `CharsetType`,
we can use like below.
```cpp
CharsetType CharsetAutoSelector::select(std::string src) {
    return (UTF8Charset().encode(src) > UTF16Charset().encode(src)) ? CharsetType::UTF16 : CharsetType::UTF8;
}
```

CharsetType is defined in Definition.h
https://github.com/euphony-io/euphony/blob/04fdaacea8b0a019f109e60440c4687b3538f1b9/euphony/src/main/cpp/core/Definitions.h#L62

Which one do you like?
if you have some question, let me know :)